### ISSUE TITLE
Detect non-lazy eval'd node attributes as defaults in resources

### ISSUE THREAD

**Issue by tas50** [chef/cookstyle#506]
### What category of cop is this?:

Please try to select as few as possible.

- [ ] ChefDeprecations - A cop that alerts on a deprecation in the Chef Infra Client
- [x] ChefCorrectness - A cop that alerts when incorrect coding behavior is being used
- [ ] ChefSharing - A cop that alerts to missing functionality necessary for sharing cookbooks with others
- [ ] ChefStyle - A cop that alerts to a style best practice
- [ ] ChefModernize - A cop that alerts when a cookbook can be simplified or modernized with new functionality
- [ ] ChefEffortless - A cop that alerts on code that must be resolved to move to the Effortless pattern

### Describe the new cop:

When writing a resource where a property has a default value that is a node attribute it should be lazy eval'd. Otherwise due to internal file parsing order things don't work the way you'd expect. Additionally, if the node attribute is not yet set it becomes nil and that fails.

### What it would trigger on:

```ruby
property :Tim, String, default: node['foo']['bar']
```

### What it would autocorrect to if applicable

```ruby
property :Tim, String, default: lazy { node['foo']['bar'] }
```

**Comment by lamont-granquist** [comment]
> compile/converge gets in the way

its got nothing to do with compile/converge though.

for core resources that line just gets parsed long before ohai even runs.

and for cookbook resources if its not an ohai resource it gets parsed before any attributes files are parsed.

so its just normal chef execution order issues.

**Comment by tyler-ball** [comment]
@lamont-granquist But won't the `lazy` block ensure the value won't be defaulted until long after ohai and the attribute files have been parsed? I'm not sure if your comment is saying "lazy won't work here" or "it isn't a compile/converge but lazy will resolve it".

**Comment by lamont-granquist** [comment]
The description of the problem isn't accurate, but solution is correct.

**Comment by tas50** [comment]
@lamont-granquist updated

**Comment by tas50** [comment]
Fixed in 6.6
